src/lime/app/Promises.hx

```diff
@@ -0,0 +1,156 @@
+
+    for (future in futures) {
+      future.onComplete(function(result) {
+        results.push(result);
+        if (results.length == futures.length) {
+          promise.complete(results);
+        }
+      });
+      future.onError(function(error) {
+        promise.error(error);
+      });
+    }
+
+    return promise.future;
+  }
+
+  /**
+   * Creates a promise which fulfills when all the input promises settle (whether success or failure).
+   * Returns an array of objects that describe the outcome of each promise.
+   * @param futures A list of Futures to resolve.
+   * @return A Future for a list of result values.
+   */
+  public static function allSettled<T>(futures:Array<Future<T>>):Future<Array<PromiseResult<T>>> {
+    var promise:Promise<Array<PromiseResult<T>>> = new Promise<Array<PromiseResult<T>>>();
+    var results:Array<PromiseResult<T>> = [];
+
+    for (future in futures) {
+      future.onComplete(function(value) {
+        results.push(PromiseResult.fulfilled(value));
+        if (results.length == futures.length) {
+          promise.complete(results);
+        }
+      });
+      future.onError(function(error) {
+        results.push(PromiseResult.rejected(value));
+        if (results.length == futures.length) {
+          promise.complete(results);
+        }
+      });
+    }
+
+    return promise.future;
+  }
+
+  /**
+   * Creates a promise which fulfills when any of the input promises resolve successfully.
+   * Returns the first fulfilled promise. If all promises reject, the promise will be rejected with the list of rejection reasons.
+   * @param futures A list of Futures to resolve.
+   * @return A Future for a result value.
+   */
+  public static function any<T>(futures:Array<Future<T>>):Future<T> {
+    var promise:Promise<T> = new Promise<T>();
+    var errors:Array<Dynamic> = [];
+
+    for (future in futures) {
+      future.onComplete(function(value) {
+        promise.complete(value);
+      });
+      future.onError(function(error) {
+        errors.push(error);
+        if (errors.length == futures.length) {
+          promise.error(errors);
+        }
+      });
+    }
+
+    return promise.future;
+  }
+
+  /**
+   * Creates a promise which fulfills when any of the input promises settle.
+   * Returns an object that describes the outcome of the first settled promise (whether success or failure).
+   * @param futures A list of Futures to resolve.
+   * @return A Future for a result value.
+   */
+  public static function race<T>(futures:Array<Future<T>>):Future<PromiseResult<T>> {
+    var promise:Promise<PromiseResult<T>> = new Promise<PromiseResult<T>>();
+    for (future in futures) {
+      future.onComplete(function(value) {
+        promise.complete(PromiseResult.fulfilled(value));
+      });
+      future.onError(function(error) {
+        promise.complete(PromiseResult.rejected(error));
+      });
+    }
+    return promise.future;
+  }
+}
+
+class PromiseResult<T> {
```

Comment:
`PromiseResult` and `PromiseState` could be combined into a single enum, with options being `Complete(value:T)` and `Error(error:Dynamic)`.

The `Pending` option isn't even used, so you'd just drop that.

Reply:
Better yet, just return the `futures:Array<Future<T>>` array that was passed into the function. That way users don't can use the API they already know instead of learning a new one.